program-runtime/src/invoke_context.rs

```diff
@@ -210,11 +212,32 @@ impl<'a> InvokeContext for ThisInvokeContext<'a> {
                 // Reentrancy not allowed unless caller is calling itself
                 return Err(InstructionError::ReentrancyNotAllowed);
             }
-        }
+            *program_id
+        } else {
+            return Err(InstructionError::UnsupportedProgramId);
```

Comment:
@jackcmay @Lichtso it looks like this new err has made master and v1.8 incompatible. On v1.8 it's possible to invoke the native loader directly from a tx in which case `program_indices` is empty and the tx can still be processed successfully.

Here is an example tx that showcases the different behavior. The tx invokes the native loader in a way that runs the code path for a 0 lamport system ix transfer with the program_id set to the native loader id: https://explorer.solana.com/tx/inspector?message=AQABAxqaPUT28ZZYeGqKkcOfzL7aLBsifPzASKGnxP4QwCbm%252Bk%252BRkBm4gzLzoCATvtlllmIDwh%252B8XErqP9gYsdFvpjUFh4S%252FFIukKC%252BwEldIiKnxU6B9rfdlwEVcmpcDgAAAAL1QHO0HDnEf1DbxkmiM2jg37QuLnqbQDF9g5Y7HZgN3AQICAAEMAgAAAAAAAAAAAAAA

It was constructed with the following JS:
```js
    // create this to copy out the ix data    
    const ix = SystemProgram.transfer({
        fromPubkey: MY_KEY.publicKey,
        toPubkey: NEW_KEY.publicKey,
        lamports: 0,
    });
    
    const tx = new Transaction().add(new TransactionInstruction({
        programId: new PublicKey("NativeLoader1111111111111111111111111111111"),
        data: ix.data,
        keys: [{
            pubkey: MY_KEY.publicKey,
            isSigner: false,
            isWritable: false,
        }, {
            pubkey: NEW_KEY.publicKey,
            isSigner: false,
            isWritable: true,
        }],
    }));
```

Reply:
As mentioned [here](https://github.com/solana-labs/solana/pull/20448/files#r752310865) I think we need to either add an empty transaction to the test corpus or feature gate and forbid them. Because they are breaking the assumption that there is at least one executable account loaded which can be the `program_id`.

Reply:
I'm in favor of feature gating and forbidding them. But we will still need to revert behavior and add a test case for this in the meantime until we enable that feature switch.

Reply:
Agreed in both, the temporary test and the plan to remove them.
It also makes life a little easier in terms of edge cases which we need to handle.

To do so:
- remove the line `debug_assert_eq!(first_instruction_account, 1);` in `process_instruction_with_program_logging()`
- add `.map_err(|_| InstructionError::UnsupportedProgramId)` behind `keyed_account_at_index(keyed_accounts, 0)?` in `instruction_processor::process_instruction()`.
- pull `InstructionError::ReentrancyNotAllowed` check out of `let program_id = if let Some(index_of_program_id) = program_indices.last() {` in `ThisInvokeContext::push()`
- and then put the `let program_id = if let Some(index_of_program_id) = program_indices.last() {` in `if self.invoke_stack.is_empty() {` without an else return error branch.
- finally, add the empty transaction test case in `program-runtime/src/invoke_context.rs`

Reply:
The change that are causing this behavior were an optimization to get the program_id in push.  We don't have to do that and instead allow empty indices and get the program id from the message instead.

Reply:
We are tracking this in a separate issue: #21346